### ISSUE TITLE
support for special matrix types?

### ISSUE THREAD

**Issue by mkborregaard** [davidavdav/NamedArrays.jl#30]
Do NamedArrays currently support special matrices - in particular I am thinking of a sparse BitMatrix?

**Comment by davidavdav** [comment]
Yes, it does seem to do so.  I'd say anything that is an `AbstractMatrix`.

**Comment by mkborregaard** [comment]
Thanks! I should be clear and say my question is does it keep the advantage of being 0/1 and sparse?
<img width="961" alt="screen shot 2016-07-13 at 18 21 24" src="https://cloud.githubusercontent.com/assets/8429802/16811039/d0e33398-4926-11e6-98c4-6977d444d37b.png">

**Comment by mkborregaard** [comment]
It looks like there is an advantage to using a sparse matrix still, though the slowdown is appreciable:
<img width="583" alt="screen shot 2016-07-14 at 12 28 50" src="https://cloud.githubusercontent.com/assets/8429802/16836647/c89d8060-49be-11e6-9c56-083782c5f38f.png">

**Comment by davidavdav** [comment]
A NamedArray has a field `array` that contains the underlying array, which can be a bitarray or a sparse array or whatever.  

I am a bit surprised that things like sum are slower, since namedarray just translated the sum() to the underlying matrix.  But maybe I have been lazy, maybe the fallback abstractarray implementation of sum is used in this case.

**Comment by mkborregaard** [comment]
Thinking about it more, I don't think so. The named sparse matrix is still faster than the dense. Could it just be overhead in assigning names to the return type (NamedArray)? In that case there is probably nothing to do about it. But thanks for responding.
Should I close the issue?

**Comment by davidavdav** [comment]
I just ran a profile, all the time is in computing and assigning new names.  I suppose especially splatting the original names into a tuple, which then get transformed to a dictionary again, is extremely inefficient. 

I would have to re-think how to make this quicker.  One of the named dimensions stays identical, I should be able to share this in the result, instead of re-computing everything clumsily.  Maybe I should introduce constructors that don't do any form of error checking.

**Comment by mkborregaard** [comment]
At least for things like `sum` that should be possible with no error checking and make a major difference performance-wise, I guess? It would be amazing if you were able to solve this performance issue, as I really like using named arrays.

**Comment by davidavdav** [comment]
I have the feeling that the real culprit is `Tuple{Dict}` I use as a holder for the indexnames.  This is a choice (I could have chosen an array instead) which makes the names more immutable.  This is so that the dictionary type is part of the NamedArray type, for the purpose of type stability.  But maybe it is defeating the purpose in this case. 

Just using the tuples as is
```
function sum1(n::NamedArrays.NamedMatrix)
       NamedArrays.NamedArray(sum(n.array,1), (Dict("sum(A)" => 1), n.dicts[2]), n.dimnames)
end
```
times at 1/3 of your namedsparse case, which is still ridiculously slow.

**Comment by davidavdav** [comment]
It looks like the assertions in the inner constructor are really the problem: leaving out the
```julia
        for (d,dict) in zip(size(array),dicts)
            Set(values(dict)) == Set(1:d) || error("Inconsistent values in dict")
        end
```
brings the timing on par with nameless array, using `sum1()` above (the normal `sum` still goes over some other constructors that do slow mapping operations).

**Comment by mkborregaard** [comment]
Fantastic! This is very very useful for me. Thanks, Michael
